Sources/Articles/Documentation.docc/2024/beginners-guide-to-protocol-buffers-and-grpc-with-swift/beginners-guide-to-protocol-buffers-and-grpc-with-swift.md

```diff
@@ -0,0 +1,382 @@
+            ]
+        ),
+    ]
+)
+```
+
+The next step is to add the `todo_messages.proto` file into the `Sources/App` directory. Additionally, create a `swift-protobuf-config.json` file to configure the Swift Protobuf plugin:
+
+```json
+{
+    "invocations": [
+        {
+            "protoFiles": [
+                "todo_messages.proto",
+            ],
+            "visibility": "internal",
+            "server": false,
+        },
+    ]
+}
+```
+
+This configuration specifies how the Swift Protobuf plugin will generate code. The `protoFiles` array includes the `todo_messages.proto` file, which will be used as the input for code generation. The `visibility` setting is defined as `internal`, and the `server` flag is set to `false`, indicating that the plugin will generate only the data models without generating any server-side code.
+
+A similar configuration is required for the gRPC Swift plugin. Add the `todo.proto` file to the source directory, and create a `grpc-swift-config.json` file with the following contents:
+
+```json
+{
+    "invocations": [
+        {
+            "protoFiles": [
+                "todo.proto",
+            ],
+            "visibility": "internal",
+            "server": true,
+            "client": false,
+            "keepMethodCasing": false,
+            "reflectionData": true
+        }
+    ]
+}
+```
+
+By using the Swift gRPC plugin, server-side Swift code is generated, and it utilizes the data types produced by the Swift Protobuf plugin.
+
+It's important to note that the Swift Protobuf plugin generates only the data structures. In contrast, the Swift gRPC plugin can generate both the server and client-side interfaces for the whole communication layer. Simply put, Protocol Buffers manage the data encoding and decoding, while gRPC uses Protocol Buffers to enable RPC communication.
+
+One caveat when using Xcode is that the plugins might need to be manually enabled. This can be done in the Report Navigator by clicking the links below the log messages. Another [issue](https://forums.swift.org/t/plugin-doesnt-have-access-to-binary-package-manager-extensible-build-tools-se-0303-and-se-0305/56038/10) is that Xcode may not access external build tools, so linking the protoc command manually may be required using the following command:
+
+```sh
+ln -snfv /opt/homebrew/bin/protoc /Applications/Xcode.app/Contents/Developer/usr/bin/protoc
+# alternatively, if you're using a beta version
+ln -snfv /opt/homebrew/bin/protoc /Applications/Xcode-beta.app/Contents/Developer/usr/bin/protoc
+```
+
+After a successful build, the generated data types and interfaces will be available, allowing the implementation of the server-side interface using the gRPC library. Below is an example of a simple actor-based implementation that uses in-memory storage and meets the `TodoService` protocol requirements:
+
+```swift
+import GRPC
+
+actor TodoService: Todos_TodoServiceAsyncProvider {
```

Comment:
Swift-protobuf v2 should be used for a new tutorial now. It's still alpha, but seems pretty solid and was recommended by cory for new apps